agent/hcp/telemetry/gauge_store_test.go

```diff
@@ -0,0 +1,87 @@
+package telemetry
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/otel/attribute"
+)
+
+func TestGaugeStore(t *testing.T) {
+	t.Parallel()
+
+	gaugeStore := NewGaugeStore()
+
+	attributes := []attribute.KeyValue{
+		{
+			Key:   attribute.Key("test_key"),
+			Value: attribute.StringValue("test_value"),
+		},
+	}
+
+	gaugeStore.Set("test", 1.23, attributes)
+
+	// Should store a new gauge.
+	val, ok := gaugeStore.LoadAndDelete("test")
+	require.True(t, ok)
+	require.Equal(t, val.Value, 1.23)
+	require.Equal(t, val.Attributes, attributes)
+
+	// Gauge with key "test" have been deleted.
+	val, ok = gaugeStore.LoadAndDelete("test")
+	require.False(t, ok)
+	require.Nil(t, val)
+
+	gaugeStore.Set("duplicate", 1.5, nil)
+	gaugeStore.Set("duplicate", 6.7, nil)
+
+	// Gauge with key "duplicate" should hold the latest (last seen) value.
+	val, ok = gaugeStore.LoadAndDelete("duplicate")
+	require.True(t, ok)
+	require.Equal(t, val.Value, 6.7)
+}
+
+func TestGaugeCallback_Failure(t *testing.T) {
+	t.Parallel()
+
+	k := "consul.raft.apply"
+	gaugeStore := NewGaugeStore()
+	gaugeStore.Set(k, 1.23, nil)
+
+	cb := gaugeStore.gaugeCallback(k)
+	ctx, cancel := context.WithCancel(context.Background())
+
+	cancel()
+	err := cb(ctx, nil)
+	require.ErrorIs(t, err, context.Canceled)
+}
+
+// TestGaugeStore_Race induces a race condition. When run with go test -race,
+// this test should pass if  implementation is concurrency safe.
+func TestGaugeStore_Race(t *testing.T) {
+	t.Parallel()
+
+	gaugeStore := NewGaugeStore()
+	wg := &sync.WaitGroup{}
+	samples := 100
+	for i := 0; i < samples; i++ {
+		k := fmt.Sprintf("consul.test.%d", i)
+		v := rand.Float64()
+		wg.Add(1)
+		go storeAndRetrieve(t, k, v, gaugeStore, wg)
+	}
+
+	wg.Wait()
+}
+
+func storeAndRetrieve(t *testing.T, k string, v float64, gaugeStore *gaugeStore, wg *sync.WaitGroup) {
+	gaugeStore.Set(k, v, nil)
+	gv, ok := gaugeStore.LoadAndDelete(k)
+	require.True(t, ok)
+	require.Equal(t, v, gv.Value)
+	wg.Done()
+}
```

Comment:
This approach is definitely more robust when testing for race conditions!

Small suggestion, I would change the goroutines to not pass the WaitGroup, and also not use a random value in order to save compute:
```suggestion
	gaugeStore := NewGaugeStore()
	wg := &sync.WaitGroup{}
	for i := 0; i < 100; i++ {
		wg.Add(1)
		key := fmt.Sprintf("consul.test.%d", i)
		value := 12.34
		go func() {
			defer wg.Done()
			
			gaugeStore.Set(key, value, nil)
			gv, ok := gaugeStore.LoadAndDelete(k)
			require.True(t, ok)
			require.Equal(t, value, gv.Value)
		}()
	}

	wg.Wait()
}
```

Reply:
I'm also unsure how the `require` package would work in this situation. I _think_ it fails the test immediately, so I'm not 100% what will happen to the `wg.Wait()` :thinking:

Reply:
Yep, it calls [`t.FailNow()`](https://github.com/stretchr/testify/blob/f36bfe3c337aa95c86f04c721acdbafb5ffb1611/require/require.go#L1825) under the hood which is not safe inside a goroutine.

I would create a buffered channel before the loop, send errors in the goroutines, and assert no errors after the call to `wg.Wait()`. Something like this should do it:
```go
func TestGaugeStore_Race(t *testing.T) {
	t.Parallel()

	gaugeStore := NewGaugeStore()
	wg := &sync.WaitGroup{}
	samples := 100
	errCh := make(chan error, samples)

	for i := 0; i < samples; i++ {
		wg.Add(1)
		key := fmt.Sprintf("consul.test.%d", i)
		value := 12.34
		go func() {
			defer wg.Done()
			gaugeStore.Set(key, value, nil)
			gv, _ := gaugeStore.LoadAndDelete(k)
			if gv.Value != value {
				errCh <- fmt.Errorf("expected value: '%s', but got: '%s' for key: '%s'", value, gv.Value, key)
			}
		}()
	}
	wg.Wait()
}

Reply:
thanks for catching this, I've fixed it and for the OTELSink test too.